compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -935,27 +938,35 @@ private boolean checkTypeAndVarCountConsistency(BLangTupleVariable varNode, BTup
             }
         }
 
-        if (tupleTypeNode.tupleTypes.size() != varNode.memberVariables.size()) {
+        if (tupleTypeNode.tupleTypes.size() != varNode.memberVariables.size()
+                || (tupleTypeNode.restType == null && varNode.restVariable != null)
+                ||  (tupleTypeNode.restType != null && varNode.restVariable == null)) {
```

Comment:
I think we should support the scenario where `tupleTypeNode.restType != null && varNode.restVariable == null`, which would allow the following:
```ballerina
public function main(string... args) {
    [int, string, string...] x = getTuple();
}

function getTuple() returns [int, string, string, string, string] {
    return [1, "A", "B", "C", "D"];
}

```